File: TESS_CppAPI_EXAMPLE/MySimulator.cpp
#include "MySimulator.h"

#include <QMessageBox>
#include <QDateTime>
#include <QDebug>

#include "ilink.h"
#include "ILane.h"
#include "IConnector.h"
#include "ILaneConnector.h"
#include "ivehicle.h"
#include "ivehicledriving.h"
#include "isignallamp.h"
#include "tessinterface.h"
#include "netinterface.h"
#include "simuinterface.h"
#include "ivehicle.h"
#include "IBusLine.h"
#include "IBusStation.h"
#include "IBusStationLine.h"
#include "UnitChange.h"
#include "AutoDynaInfo.h"

MySimulator::MySimulator()
	:mrSquareVehiCount(28), mrSpeedOfPlane(-1), mrStartMSecs(0)
{

}

//开始仿真前的做准备
void MySimulator::beforeStart(bool &keepOn) {
	//当前路网名
	QString tmpNetPath = gpTessInterface->netInterface()->netFilePath();
	if (tmpNetPath != mNetPath)
	{
		mNetPath = tmpNetPath;
		mSimuCount = 0;
	}
	gpTessInterface->simuInterface()->setAcceMultiples(1);

	//开始仿真前修改公交线路站点部分参数
	QList<IBusLine*> lBusLine = gpTessInterface->netInterface()->buslines();
	if (!lBusLine.isEmpty())
	{
		IBusLine* pLine = lBusLine.first();
		long id = pLine->id();
		QList<ILink*> lLink = pLine->links();
		QList<IBusStationLine*> lStationLine = pLine->stationLines();
		if (!lStationLine.isEmpty())
		{
			IBusStationLine* pStationLine = lStationLine.first();
			pStationLine->setBusParkingTime(20);

		}
	}

	//如果return true且keepOn为false，则仿真不被执行，可以偿试将keepOn设为false，观察是否执行仿真
	keepOn = true;
	//如果返回false，不判断keepOn的值，开始仿真
}

//仿真起动后的处理
void MySimulator::afterStart() {
	//像素比，米/像素
	//mrScale = gpTessInterface->netInterface()->sceneScale();
}

//仿真结束后的处理
void MySimulator::afterStop() {
	//QDateTime dt = QDateTime::fromMSecsSinceEpoch(mrStartMSecs);
	//QString dtStr = dt.toString("yyyy年MM月dd日 hh时mm分ss秒");
	//QString startDtStr = QString("仿真开始现实时间:").append(dtStr);

	//dtStr = QDateTime::currentDateTime().toString("yyyy年MM月dd日 hh时mm分ss秒");
	//QString endDtStr = QString("仿真结束现实时间:").append(dtStr);

	//QMessageBox::information(nullptr, QString(), QString().append(startDtStr).append("\n").append(endDtStr));

	mSimuCount++;
	QString netfilePath = gpTessInterface->netInterface()->netFilePath();
	if (netfilePath.contains("深圳南海大道创业路交叉口CFI方案") && mSimuCount < 3)
	{
		emit(forReStartSimu());
	}

}

//初始车辆, laneNumber:车道，从0开始；dist，距起点距离，单位像素；speed：车速，像素/秒
//bool MySimulator::initVehicle(IVehicle *pIVehicle, int laneNumber, qreal dist, qreal speed) {
//初始车辆
void MySimulator::initVehicle(IVehicle *pIVehicle) {
	//======设置部分方法被TESSNG调用频次，参数"__custsimubysteps"设为true时下列设置起作用======
	//每20个计算周期调用1次
	pIVehicle->setSteps_reCalcToLeftFreely(20);
	//每20个计算周期调用1次
	pIVehicle->setSteps_reCalcToRightFreely(20);
	//每个计算周期都调用
	pIVehicle->setSteps_reCalcdesirSpeed(1);
	//每个计算周期都调用
	pIVehicle->setSteps_reSetSpeed(1);

	long tmpId = pIVehicle->id() % 100000;
	QString roadName = pIVehicle->roadName();
	long roadId = pIVehicle->roadId();
	if (roadName == QString("曹安公路")) {
		//飞机
		if (tmpId == 1) {
			pIVehicle->setVehiType(12);
			pIVehicle->initLane(3, m2p(105), 0);
		}
		//工程车
		else if (tmpId >= 2 && tmpId <= 8) {
			pIVehicle->setVehiType(8);
			pIVehicle->initLane((tmpId - 2) % 7, m2p(80), 0);
		}
		//消防车
		else if (tmpId >= 9 && tmpId <= 15) {
			pIVehicle->setVehiType(9);
			pIVehicle->initLane((tmpId - 2) % 7, m2p(65), 0);
		}
		//救护车
		else if (tmpId >= 16 && tmpId <= 22) {
			pIVehicle->setVehiType(10);
			pIVehicle->initLane((tmpId - 2) % 7, m2p(50), 0);
		}
		//最后两队列小车
		else if (tmpId == 23) {
			pIVehicle->setVehiType(1);
			pIVehicle->initLane(1, m2p(35), 0);
		}
		else if (tmpId == 24) {
			pIVehicle->setVehiType(1);
			pIVehicle->initLane(5, m2p(35), 0);
		}
		else if (tmpId == 25) {
			pIVehicle->setVehiType(1);
			pIVehicle->initLane(1, m2p(20), 0);
		}
		else if (tmpId == 26) {
			pIVehicle->setVehiType(1);
			pIVehicle->initLane(5, m2p(20), 0);
		}
		else if (tmpId == 27) {
			pIVehicle->setVehiType(1);
			pIVehicle->initLane(1, (5), 0);
		}
		else if (tmpId == 28) {
			pIVehicle->setVehiType(1);
			pIVehicle->initLane(5, (5), 0);
		}
		if (tmpId >= 23 && tmpId <= 28) {
			pIVehicle->setLength(m2p(4.5), true);
		}
	}
}

//计算加速度
bool MySimulator::calcAcce(IVehicle *pIVehicle, qreal &acce)
{
	AutoDynaInfo *pDynaInfo = (AutoDynaInfo*)pIVehicle->dynaInfo();
	//如果是自动驾驶车辆，将计数加1，如果车速大于10m/s，将加速度设为-2m/s^2，如果车速小于1m/s，将加速度设为2m/s
	if (pDynaInfo)
	{
		pDynaInfo->mrCalcCount++;
		if(pIVehicle->vehiDistFront() < m2p(5))
		{
			//前车距小于5米，让TESSNG计算加速度
			return false;
		}
		else if (pIVehicle->currSpeed() > m2p(10))
		{
			acce = m2p(-2);
		}
		else if (pIVehicle->currSpeed() < m2p(1))
		{
			acce = m2p(2);
		}
		return true;
	}
	return false;
}

//重新计算期望速度
bool MySimulator::reCalcdesirSpeed(IVehicle *pIVehicle, qreal &inOutDesirSpeed) {
	long tmpId = pIVehicle->id() % 100000;
	QString roadName = pIVehicle->roadName();
	if (roadName == QString("曹安公路")) {
		if (tmpId <= mrSquareVehiCount) {
			//当前已仿真时间
			long simuTime = gpTessInterface->simuInterface()->simuTimeIntervalWithAcceMutiples();
			if (simuTime < 5 * 1000) {
				inOutDesirSpeed =  0;
			} else if (simuTime < 10 * 1000) {
				inOutDesirSpeed = m2p(20 / 3.6);
			}
			else {
				inOutDesirSpeed = m2p(40 / 3.6);
			}
			return true;
		}
	}
	return false;
}

//重新设置跟驰的安全时距及安全距离， inOutSi：安全时距，inOutSd: 安全距离
bool MySimulator::reSetFollowingParam(IVehicle *pIVehicle, qreal &inOutSi, qreal &inOutSd) {
	QString roadName = pIVehicle->roadName();
	if (roadName == QString("连接段2")) {
		inOutSd = m2p(30);
		return true;
	}
	return false;
}

QList<Online::DispatchInterval> MySimulator::calcDynaDispatchParameters()
{
	QList<Online::DispatchInterval> lDispatchInterval;
	long simuTime = gpTessInterface->simuInterface()->simuTimeIntervalWithAcceMutiples();
	if (simuTime == 30 * 1000)
	{
		QTime t = QTime::currentTime();
		long currSecs = t.hour() * 3600 + t.minute() * 60 + t.second();
		Online::DispatchInterval di = Online::DispatchInterval();
		di.dispatchId = 1;
		di.fromTime = currSecs;
		di.toTime = di.fromTime + 300 - 1;
		di.vehiCount = 300;
		di.mlVehicleConsDetail << Online::VehicleComposition(1, 90) << Online::VehicleComposition(2, 10);
		lDispatchInterval << di;
	}
	return lDispatchInterval;
}

QList<Online::DecipointFlowRatioByInterval> MySimulator::calcDynaFlowRatioParameters()
{
	QList<Online::DecipointFlowRatioByInterval> lDecipointFLowRatioByInterval;
	long batchNumber = gpTessInterface->simuInterface()->batchNumber();
	if (batchNumber == 100)
	{
		Online::DecipointFlowRatioByInterval dfi = Online::DecipointFlowRatioByInterval();
		dfi.deciPointID = 1;
		dfi.startDateTime = 1;
		dfi.endDateTime = 84000;
		Online::RoutingFlowRatio rfr1 = Online::RoutingFlowRatio(1, 10);
		Online::RoutingFlowRatio rfr2 = Online::RoutingFlowRatio(2, 1);
		dfi.mlRoutingFlowRatio << rfr1;
		dfi.mlRoutingFlowRatio << rfr2;
		lDecipointFLowRatioByInterval << dfi;
	}
	return lDecipointFLowRatioByInterval;
}

//如果在路段上，重新设置车辆加速度，注意：重新设置的加速度导致的车速不受道路限速影响
bool MySimulator::reSetAcce(IVehicle *pIVehicle, qreal &inOutAcce) {
	QString roadName = pIVehicle->roadName();
	if (roadName == QString("连接段1")) {
		//如果车速大于20km/h，加速度设为-5m/s/s
		if (pIVehicle->currSpeed() > m2p(20 / 3.6)) {
			inOutAcce = m2p(-5);
			return true;
		}
		//如果车速大于10km/h，加速度设为-1m/s/s
		else if (pIVehicle->currSpeed() > m2p(20 / 3.6)) {
			inOutAcce = m2p(-1);
			return true;
		}
	}
	return false;
}

//重新设置速度
bool MySimulator::reSetSpeed(IVehicle *pIVehicle, qreal &inOutSpeed) {
	long tmpId = pIVehicle->id() % 100000;
	QString roadName = pIVehicle->roadName();
	if (roadName == QString("曹安公路")) {
		if (tmpId == 1) {
			mrSpeedOfPlane = pIVehicle->currSpeed();
		}
		else if (tmpId >= 2 && tmpId <= mrSquareVehiCount && mrSpeedOfPlane >= 0) {
			inOutSpeed = mrSpeedOfPlane;
			return true;
		}
	}
	long linkid = pIVehicle->startLink()->id();
	long startDt = pIVehicle->startSimuTime();
	return false;
}

//计算是否要左自由变道
bool MySimulator::reCalcToLeftFreely(IVehicle *pIVehicle) {
	//车辆到路段或连接终点距离小于20米不变道
	if (pIVehicle->vehicleDriving()->distToEndpoint() - pIVehicle->length() / 2 < m2p(20)) {
		return false;
	}

	long tmpId = pIVehicle->id() % 100000;
	QString roadName = pIVehicle->roadName();
	if (roadName == QString("曹安公路")) {
		if (tmpId >= 23 && tmpId <= 28) {
			int laneNumber = pIVehicle->vehicleDriving()->laneNumber();
			if (laneNumber == 1 || laneNumber == 4) {
				return true;
			}
		}
	}
	return false;
}

//计算是否要右自由变道
bool MySimulator::reCalcToRightFreely(IVehicle *pIVehicle) {
	long tmpId = pIVehicle->id() % 100000;
	//车辆到路段或连接终点距离小于20米不变道
	if (pIVehicle->vehicleDriving()->distToEndpoint() - pIVehicle->length() / 2 < m2p(20)) {
		return false;
	}

	if (tmpId == 27)
	{
		qDebug();
	}

	QString roadName = pIVehicle->roadName();
	if (roadName == QString("曹安公路")) {
		if (tmpId >= 23 && tmpId <= 28) {
			int laneNumber = pIVehicle->vehicleDriving()->laneNumber();
			if (laneNumber == 2 || laneNumber == 5) {
				return true;
			}
		}
	}
	return false;
}

//计算信号灯颜色
bool MySimulator::calcLampColor(ISignalLamp *pSignalLamp) {
	if (pSignalLamp->id() == 5)
	{
		pSignalLamp->setLampColor("R");
		return true;
	}
	long batchNumber = gpTessInterface->simuInterface()->batchNumber();
	//每计算200批次进行一次信号灯色的变化
	//if ((batchNumber / 200) % 2 == 0) {
	//	pSignalLamp->setLampColor(Qt::green);
	//}
	//else {
	//	pSignalLamp->setLampColor(Qt::red);
	//}
	//if (pSignalLamp->id() == 1)
	//{
	//	pSignalLamp->setLampColor(Qt::red);
	//}
	//else if (pSignalLamp->id() == 2) {
	//	pSignalLamp->setLampColor(Qt::green);
	//}
	//else {
	//	pSignalLamp->setLampColor(Qt::gray);
	//}
	return false;
}

//一个批次计算后的处理
void MySimulator::afterOneStep() {
	//============以下是获取一些仿真过程数据的方法============
	//当前仿真计算批次
	long batchNum = gpTessInterface->simuInterface()->batchNumber();
	//当前已仿真时间，单位：毫秒
	long simuTime = gpTessInterface->simuInterface()->simuTimeIntervalWithAcceMutiples();
	if (batchNum == 20 * 20)
	{
		gpTessInterface->simuInterface()->setAcceMultiples(100);
	}
	//开始仿真的现实时间
	qint64 startRealtime = gpTessInterface->simuInterface()->startMSecsSinceEpoch();
	//所有在运行车辆
	QList<IVehicle*> lAllVehi = gpTessInterface->simuInterface()->allVehiStarted();
	//ID为1的路段上所有已启动车辆
	QList<IVehicle*> lVehi = gpTessInterface->simuInterface()->vehisInLink(1);
	//车辆轨迹
	QList<Online::VehicleStatus> lVehiStatus = gpTessInterface->simuInterface()->getVehisStatus();
	
	//车辆数据采集信息
	QList<Online::VehiInfoCollected> lOutVehiInfo = gpTessInterface->simuInterface()->getVehisInfoCollected();
	//采集器集计信息
	QList<Online::VehiInfoAggregated> lOutVehiInfoAggre = gpTessInterface->simuInterface()->getVehisInfoAggregated();
	//车辆排队信息
	QList<Online::VehiQueueCounted> lOutVehiQueue = gpTessInterface->simuInterface()->getVehisQueueCounted();
	//车辆排队集计数据
	QList<Online::VehiQueueAggregated> lOutVehiQueueAggre = gpTessInterface->simuInterface()->getVehisQueueAggregated();
	//行程序时间采集信息
	QList<Online::VehiTravelDetected> lOutVehiTravel = gpTessInterface->simuInterface()->getVehisTravelDetected();
	//行程时间集计数据
	QList<Online::VehiTravelAggregated> lOutVehiTravelAggre = gpTessInterface->simuInterface()->getVehisTravelAggregated();
	//当前延误
	long delay = gpTessInterface->simuInterface()->delayTimeOnBatchNumber(batchNum);

	mRunInfo = QString("路段数：%0\n运行车辆数：%1\n仿真时间：%2(毫秒)").arg(gpTessInterface->netInterface()->linkCount()).arg(lAllVehi.size()).arg(simuTime);

	if (batchNum % 20 == 0) {
		QString info = mRunInfo;
		emit(forRunInfo(info));
	}

	//动态发车，不通过发车点
	if (batchNum % 50 == 1)
	{
		QString color = QString("#%1%2%3").arg(QString::number(qrand() % 256, 16)).arg(QString::number(qrand() % 256, 16)).arg(QString::number(qrand() % 256, 16));

		//路段上发车
		Online::DynaVehiParam dvp;
		dvp.vehiTypeCode = qrand() % 4 + 1;
		dvp.roadId = 6;
		dvp.laneNumber = qrand() % 3;
		dvp.dist = 50;
		dvp.speed = 20;
		dvp.color = color;
		IVehicle *pIVehicle1 = gpTessInterface->simuInterface()->createGVehicle(dvp);
		if (pIVehicle1)
		{
			AutoDynaInfo *pInfo = new AutoDynaInfo();
			pInfo->mName = QString("自动驾驶_%1").arg(pIVehicle1->id());
			pIVehicle1->setDynaInfo(pInfo);
		}

		//连接段上发车
		Online::DynaVehiParam dvp2;
		dvp2.vehiTypeCode = qrand() % 4 + 1;
		dvp2.roadId = 3;
		dvp2.laneNumber = qrand() % 3;
		dvp2.toLaneNumber = dvp2.laneNumber; //默认为-1，如果大于等于0, 在连接段上发车
		dvp2.dist = 50;
		dvp2.speed = 20;
		dvp2.color = color;
		IVehicle *pIVehicle2 = gpTessInterface->simuInterface()->createGVehicle(dvp2);
		if (pIVehicle2)
		{
			AutoDynaInfo *pInfo = new AutoDynaInfo();
			pInfo->mName = QString("自动驾驶_%1").arg(pIVehicle2->id());
			pIVehicle2->setDynaInfo(pInfo);
		}
	}

	if (simuTime >= 300 * 1000) {
		QString netfilePath = gpTessInterface->netInterface()->netFilePath();
		if (netfilePath.contains("深圳南海大道创业路交叉口CFI方案") && mSimuCount < 3)
		{
			emit(forStopSimu());
		}
	}
}

//绘制车辆，绘制前将车辆对象旋转指定角度
bool MySimulator::paintVehicle(IVehicle *pIVehicle, QPainter *painter)
{
	return false;
}

MySimulator::~MySimulator(){
}

File: TESS_PythonAPI_demo/4交通事件模拟/单车道事故/MySimulator.py
from PySide2.QtCore import *
from shiboken2.shiboken2 import wrapInstance

from Tessng import TessInterface, SimuInterface, PyCustomerSimulator, IVehicle, ILink
from Tessng import m2p, p2m, tngIFace, tngPlugin
from Tessng import *
import random
from datetime import datetime

# 用户插件子类，代表用户自定义与仿真相关的实现逻辑，继承自PyCustomerSimulator
#     多重继承中的父类QObject，在此目的是要能够自定义信号signlRunInfo
class MySimulator(QObject, PyCustomerSimulator, SimuInterface):
    signalRunInfo = Signal(str)
    def __init__(self):
        QObject.__init__(self)
        PyCustomerSimulator.__init__(self)


    # 过载的父类方法， 初始化车辆，在车辆启动上路时被TESS NG调用一次
    def initVehicle(self, vehi):
        # 车辆ID，不含首位数，首位数与车辆来源有关，如发车点、公交线路
        global tmpId
        global simuTime
        iface = tngIFace()
        simuIFace = iface.simuInterface()
        simuTime = simuIFace.simuTimeIntervalWithAcceMutiples()
        #     roadName = vehi.roadName()
        tmpId = vehi.id() % 100000
        # 车辆所在路段名或连接段名
        roadName = vehi.roadName()

        # 车辆所在路段ID或连接段ID
        roadId = vehi.roadId()
        if roadName == '路段1':
            vehi.initSpeed(m2p(150))

            if tmpId == 1:
                vehi.setVehiType(2)

            elif tmpId >= 2 and tmpId <= 4:
                vehi.setVehiType(8)

            elif tmpId >= 5 and tmpId <= 6:
                vehi.setVehiType(9)

            elif tmpId >= 7 and tmpId <= 9:
                vehi.setVehiType(10)

            elif 12 >= tmpId >= 9:
                vehi.setVehiType(8)

            elif 22 >= tmpId >= 12:
                vehi.setVehiType(1)

            elif 24 >= tmpId >= 22:
                vehi.setVehiType(9)

            elif 50 >= tmpId >= 25:
                vehi.setVehiType(1)

            elif tmpId >= 54 and tmpId <= 51:
                vehi.setVehiType(10)

            elif tmpId >= 55 and tmpId <= 56:
                vehi.setVehiType(8)

            else:
                vehi.setVehiType(1)

        return True

    # #前方车辆停车后，后方车辆期望速度设置为10
    # def ref_reCalcdesirSpeed(self, vehi, ref_desirSpeed):
    #     tmpId = vehi.id() % 100000
    #     roadName = vehi.roadName()
    #     if roadName == '路段1':
    #         iface = tngIFace()
    #         simuIFace = iface.simuInterface()
    #         simuTime = simuIFace.simuTimeIntervalWithAcceMutiples()
    #         if simuTime >= 28000:
    #             ref_desirSpeed.value = m2p(10)
    #             return True
    #     return False


    #设置连接段1的安全距离
    def ref_reSetFollowingParam(self, vehi, ref_inOutSi, ref_inOutSd):
        roadName = vehi.roadName()
        if roadName == "连接段1":
            ref_inOutSd.value = m2p(5)
            return True
        return False


    #     # 过载的父类方法，计算是否要左自由变道
    #     # vehi:车辆
    #     # return结果，True：变道、False：不变道
    # def reCalcToLeftFreely(self, vehi):
    #     # 车辆到路段终点距离小于20米不变道
    #     # 车辆到路段终点距离小于20米不变道
    #     if vehi.vehicleDriving().distToEndpoint() - vehi.length() / 2 < m2p(20):
    #         return False
    #     roadName = vehi.roadName()
    #     if roadName == "路段1":
    #         if simuTime >= 26000 and vehi.vehicleTypeCode() < 2:
    #             return True
    #     return False
    #
    # 
    #     # 过载的父类方法，计算是否要右自由变道
    #     # vehi:车辆
    #     # return结果，True：变道、False：不变道
    # def reCalcToRightFreely(self, vehi):
    #     # 车辆到路段终点距离小于20米不变道
    #     if vehi.vehicleDriving().distToEndpoint() - vehi.length() / 2 < m2p(20):
    #         return False
    #     roadName = vehi.roadName()
    #     if roadName == "路段1":
    #         if simuTime >= 26000 and vehi.vehicleTypeCode() > 2:
    #             return True
    #     return False

    #车辆故障停车
    def ref_reSetSpeed(self, vehi, ref_inOutSpeed):
        link = wrapInstance(vehi.road().__int__(), ILink)
        tmpId = vehi.id() % 100000

        if tmpId == 11 and simuTime >= 78000 and simuTime <= 108000:
            # print(tmpId, vehi.vehicleTypeCode())
            # if tmpId == 11 and vehi.vehicleTypeCode() == 2:
            if vehi.vehicleTypeCode() != 12:
                ref_inOutSpeed.value = m2p(0)

                return True
        return False

    # 过载父类方法， 计算车辆当前限制车道序号列表
    def calcLimitedLaneNumber(self, vehi):
        # 如果当前车辆在路段上，且路段ID等于3，则小车走内侧，大车走外侧
        if vehi.vehicleTypeCode() == 1:
            if vehi.roadIsLink():
                # IVehicle.road()方法获取的是车辆当前所路段或连接段的void指针，需要将它转换成路段或连接段
                link = wrapInstance(vehi.road().__int__(), ILink)
                if link is not None and link.id() == 3 and simuTime >= 25000:
                    # 小车走内侧，大车走外侧，设长度小于6米为小车
                    return []
        return []



    # 过载父类方法，动态修改决策点不同路径流量比
    def calcDynaFlowRatioParameters(self):
        # TESSNG 顶层接口
        iface = tngIFace()
        # 当前仿真计算批次
        batchNum = iface.simuInterface().batchNumber()
        # 在计算第20批次时修改某决策点各路径流量比
        if batchNum == 20:
            # 一个决策点某个时段各路径车辆分配比
            dfi = TESSNG.DecipointFlowRatioByInterval()
            #决策点编号
            dfi.deciPointID = 5
            #起始时间 单位秒
            dfi.startDateTime = 1
            #结束时间 单位秒
            dfi.endDateTime = 84000
            rfr1 = TESSNG.RoutingFlowRatio(10, 3)
            rfr2 = TESSNG.RoutingFlowRatio(11, 4)
            rfr3 = TESSNG.RoutingFlowRatio(12, 3)
            dfi.mlRoutingFlowRatio = [rfr1, rfr2, rfr3]
            return [dfi]
        return []


    def isStopDriving(self, vehi):
        if vehi.roadId() == 2:
            # 车头到当前路段或连接段终点距离
            dist = vehi.vehicleDriving().distToEndpoint(True)
            # 如果距终点距离小于100米，车辆停止运行退出路网
            if dist < m2p(5):
                return False
        return False

    # 过载的父类方法，TESS NG 在每个计算周期结束后调用此方法，大量用户逻辑在此实现，注意耗时大的计算要尽可能优化，否则影响运行效率
    def afterOneStep(self):
        #= == == == == == =以下是获取一些仿真过程数据的方法 == == == == == ==
        # TESSNG 顶层接口
        iface = tngIFace()
        # TESSNG 仿真子接口
        simuiface = iface.simuInterface()
        # TESSNG 路网子接口
        netiface = iface.netInterface()
        # 当前仿真计算批次
        batchNum = simuiface.batchNumber()
        # 当前已仿真时间，单位：毫秒
        simuTime = simuiface.simuTimeIntervalWithAcceMutiples()
        # 开始仿真的现实时间
        startRealtime = simuiface.startMSecsSinceEpoch()
        # 当前正在运行车辆列表
        lAllVehi = simuiface.allVehiStarted()
        # 打印当前在运行车辆ID列表
        # print([item.id() for item in lAllVehi])
        # 当前在ID为1的路段上车辆
        lVehis = simuiface.vehisInLink(1)
        #每20个计算批次做一次小计，将结果通过信号发送出云
        if batchNum % 20 == 0:
            strLinkCount = str(netiface.linkCount())
            strVehiCount = str(len(lAllVehi))
            strSimuTime = str(simuTime)
            runInfo = f"路段数：{strLinkCount}\n运行车辆数：{strVehiCount}\n仿真时间：{strSimuTime}(毫秒)"
            self.signalRunInfo.emit(runInfo)

        # 动态发车，不通过发车点发送，直接在路段和连接段中间某位置创建并发送，每50个计算批次发送一次
        if batchNum % 50 == 1:
            r = hex(256 + random.randint(0,256))[3:].upper()
            g = hex(256 + random.randint(0,256))[3:].upper()
            b = hex(256 + random.randint(0,256))[3:].upper()
            color = f"#{r}{g}{b}"
            # 路段上发车
            dvp = TESSNG.DynaVehiParam()
            dvp.vehiTypeCode = random.randint(0, 4) + 1
            dvp.roadId = 6
            dvp.laneNumber = random.randint(0, 3)
            dvp.dist = 50
            dvp.speed = 20
            dvp.color = color
            vehi1 = simuiface.createGVehicle(dvp)
            if vehi1 != None:
                pass
    #
    #         # 连接段上发车
    #         dvp2 = TESSNG.DynaVehiParam()
    #         dvp2.vehiTypeCode = random.randint(0, 4) + 1
    #         dvp2.roadId = 3
    #         dvp2.laneNumber = random.randint(0, 3)
    #         dvp2.toLaneNumber = dvp2.laneNumber # 默认为 - 1，如果大于等于0, 在连接段上发车
    #         dvp2.dist = 50
    #         dvp2.speed = 20
    #         dvp2.color = color
    #         vehi2 = simuiface.createGVehicle(dvp2)
    #         if vehi2 != None:
    #             pass

        # 获取车辆状态，含位置
        lVehiStatus = simuiface.getVehisStatus()
        #print("车辆位置：", [(status.vehiId, status.mPoint.x(), status.mPoint.y()) for status in lVehiStatus])
        # 信号灯组相位颜色
        lPhoneColor = simuiface.getSignalPhasesColor()
        #print("信号灯组相位颜色", [(pcolor.signalGroupId, pcolor.phaseNumber, pcolor.color, pcolor.mrIntervalSetted, pcolor.mrIntervalByNow) for pcolor in lPhoneColor])
        # 获取当前仿真时间完成穿越采集器的所有车辆信息
        lVehiInfo = simuiface.getVehisInfoCollected()
        #if len(lVehiInfo) > 0:
        #    print("车辆信息采集器采集信息：", [(vinfo.collectorId, vinfo.vehiId) for vinfo in lVehiInfo])
        # 获取最近集计时间段内采集器采集的所有车辆集计信息
        lVehisInfoAggr = simuiface.getVehisInfoAggregated()
        #if len(lVehisInfoAggr) > 0:
        #    print("车辆信息采集集计数据：", [(vinfo.collectorId, vinfo.vehiCount) for vinfo in lVehisInfoAggr])
        # 获取当前仿真时间排队计数器计数的车辆排队信息
        lVehiQueue = simuiface.getVehisQueueCounted()
        #if len(lVehiQueue) > 0:
        #    print("车辆排队计数器计数：", [(vq.counterId, vq.queueLength) for vq in lVehiQueue])
        # 获取最近集计时间段内排队计数器集计数据
        lVehiQueueAggr = simuiface.getVehisQueueAggregated()
        #if len(lVehiQueueAggr) > 0:
        #    print("车辆排队集计数据：", [(vqAggr.counterId, vqAggr.avgQueueLength) for vqAggr in lVehiQueueAggr])
        # 获取当前仿真时间行程时间检测器完成的行程时间检测信息
        lVehiTravel = simuiface.getVehisTravelDetected()
        #if len(lVehiTravel) > 0:
        #    print("车辆行程时间检测信息：", [(vtrav.detectedId, vtrav.travelDistance) for vtrav in lVehiTravel])
        # 获取最近集计时间段内行程时间检测器集计数据
        lVehiTravAggr = simuiface.getVehisTravelAggregated()
        #if len(lVehiTravAggr) > 0:
        #    print("车辆行程时间集计数据：", [(vTravAggr.detectedId, vTravAggr.vehiCount, vTravAggr.avgTravelDistance) for vTravAggr in lVehiTravAggr])
























